Add questions and notes for pr20827 (#440)
bitcoin-core-review-club/website: _posts/2021-12-29-#20827.md

commit:
---

## Notes

- When pruning is enabled in Bitcoin Core, data about old blocks is deleted to limit disk space usage.
Users can configure a pruning target defining how much disk space to use for block and undo data.

> The minimum target is 550MB.

Reviewer: Suggestion - mention and/or link to `-prune` option

- Node operators can control the buffer size with the `-dbcache=<size in MB>` argument.
A larger buffer takes up more system memory but takes longer to fill and thus requires fewer flushes.
This speeds up the initial block download (IBD).

- Pruning is a reason for us to flush the dbcache regardless of its memory usage.
The maximum configured dbcache size is often not reached.

- This PR changes the pruning behavior.
Previously, we'd prune just enough files for us to be able to continue the IBD.
We now aggressively prune all prunable files enabling us to continue with IBD without having to prune again too soon.
Fewer prunes also mean fewer dbcache flushes, potentially speeding IBD for pruned nodes up.
Higher dbcache sizes can be reached before the dbcache is flushed.

<p style="text-align:center"><a href="../assets/img/20827-pruning.drawio.png"><img src="../assets/img/20827-pruning.drawio.png" style="max-width:100%; height:auto"></a></p>

- PR [#12404](https://github.com/bitcoin/bitcoin/pull/12404) attempted aggressive pruning too, but was closed in favor of PR [#11658](https://github.com/bitcoin/bitcoin/pull/11658).
PR #11658 added 10% of the prune target to the `nBuffer`.

> This is being overwritten by this PR.

Reviewer: ```suggestion
This is being overwritten by PR #20827.
```

## Questions

> 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?

Reviewer: Doesn't need to be a question here, but I recommend asking "what does this PR do" during the meeting, especially since the diff is so small.

> 2. Where in the code do we check if we need to prune old block data?

Reviewer: Perhaps add a hint for what to grep for?

3. What is removed during pruning and under which conditions? What is not pruned?

4. What is the variable `nBuffer` in [`BlockManager:FindFilesToPrune()`](https://github.com/bitcoin/bitcoin/blob/24f3936337de3afb4fa56efc83009e2527d22df0/src/validation.cpp#L3622) being used for? How large is the buffer (in MB)?

5. The PR assumes 1MB for `average_block_size`. How accurate does this assumption have to be?

> 6. The PR description mentions IBD speed improvements for pruned nodes. Which metrics can we measure to benchmark the improvement? With which prune targets and dbcache sizes should we test?

Reviewer: ```suggestion
6. The PR description mentions IBD speed improvements for pruned nodes. What can we measure to benchmark the improvement? With which prune targets and dbcache sizes should we test?
```

> 7. Edge case: Is agressivly pruning during IBD a problem in-case there are longer forks in the chain?

Reviewer: ```suggestion
7. Edge case: Is agressively pruning during IBD a problem if there are longer forks in the chain?
```
